python/hail/api2/matrixtable.py

```diff
@@ -263,55 +340,181 @@ def _jvdf(self):
 
     @property
     def global_schema(self):
+        """The schema of global fields in the matrix.
+
+        Returns
+        -------
+        :class:`hail.expr.TStruct`
+            Global schema.
+        """
         if self._global_schema is None:
             self._global_schema = Type._from_java(self._jvds.globalSignature())
         return self._global_schema
 
     @property
     def colkey_schema(self):
+        """The schema of the column key.
+
+        Returns
+        -------
+        :class:`hail.expr.Type`
+             Column key schema.
+        """
         if self._colkey_schema is None:
             self._colkey_schema = Type._from_java(self._jvds.sSignature())
         return self._colkey_schema
 
     @property
     def col_schema(self):
+        """The schema of column-indexed fields in the matrix.
+
+        Returns
+        -------
+        :class:`hail.expr.TStruct`
+             Column schema.
+        """
         if self._sa_schema is None:
             self._sa_schema = Type._from_java(self._jvds.saSignature())
         return self._sa_schema
 
     @property
     def rowkey_schema(self):
+        """The schema of the row key.
+
+        Returns
+        -------
+        :class:`hail.expr.Type`
+             Row key schema.
+        """
         if self._rowkey_schema is None:
             self._rowkey_schema = Type._from_java(self._jvds.vSignature())
         return self._rowkey_schema
 
     @property
     def row_schema(self):
+        """The schema of row-indexed fields in the matrix.
+
+        Returns
+        -------
+        :class:`hail.expr.TStruct`
+             Row schema.
+        """
         if self._va_schema is None:
             self._va_schema = Type._from_java(self._jvds.vaSignature())
         return self._va_schema
 
     @property
     def entry_schema(self):
+        """The schema of row-and-column-indexed fields in the matrix.
+
+        Returns
+        -------
+        :class:`hail.expr.TStruct`
+             Entry schema.
+        """
         if self._genotype_schema is None:
             self._genotype_schema = Type._from_java(self._jvds.genotypeSignature())
         return self._genotype_schema
 
     @handle_py4j
     def annotate_globals(self, **named_exprs):
+        """Create new global fields by name.
+
+        Examples
+        --------
+        Add two global fields:
+
+        >>> pops_1kg = {'EUR', 'AFR', 'EAS', 'SAS', 'AMR'}
+        >>> dataset_result = dataset.annotate_globals(pops_in_1kg = pops_1kg,
+        ...                                           gene_list = ['SHH', 'SCN1A', 'SPTA1', 'DISC1'])
+
+        Add global fields from another table and matrix table:
+
+        >>> dataset_result = dataset.annotate_globals(thing1 = dataset2[:, :].global_field,
+        ...                                           thing2 = table1[:].global_field)
```

Comment:
Are the slices here necessary?  Should `annotate_globals` be able to take a indexless expression even if it is from another object?

Reply:
this is a longer conversation. I think yes, we do need the `[:]`. This permits us to have drastically simplified expression unification logic, and probably nicer error messages when the user is doing something wrong.